Automatically track JDK of eclipse.jdt.ls

mavaddat opened the pull request:
I created this action and script by closely scrutinizing the logic of the pull requests for this and the eclipse-jdtls/eclipse.jdt.ls repos: 

- https://github.com/redhat-developer/vscode-java/pull/3021/
- https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/2557
- https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/2586

This led me to generalize the workflow using the [very convenient API endpoints for the eclipse-jdtls CI pipeline](https://ci.eclipse.org/ls/job/jdt-ls-master/).

I tested the scripts on my local machine, but of course, the resources and tokens on the runner are different. Needs testing.

mavaddat commented:
I have tested this action on my fork. I believe `peter-evans/create-pull-request@v5` action requires that the `Allow GitHub Actions to create and approve pull requests` setting be enabled on the repository. 

![Actions settings · vscode-java Screenshot](https://github.com/redhat-developer/vscode-java/assets/5055400/44d5dfce-9c17-480a-b2a8-2ed52d7041ae)

This setting can be found in a repository's settings under `Actions > General > Workflow permissions`. For repositories belonging to an organization, this setting can be managed by admins in organization settings under `Actions > General > Workflow permissions`. This is necessary because the action needs the appropriate permissions to create a pull request on our behalf.

mavaddat commented on .github/workflows/bump-jdk.yml:
```diff
@@ -0,0 +1,46 @@
+
+      - name: Branch and push changes
+        if: ${{ success() && env.latest_jdk != '' }}
+        uses: peter-evans/create-pull-request@v5
```
I had looked at that, but I thought it required setting up `gh` GitHub CLI. I see now that the `gh` CLI is preinstalled on all GitHub-hosted runners, so there is no need to set it up first. However, in order to use the `gh` CLI in a GitHub action, you need to set an environment variable called `GH_TOKEN` to a token with the required scopes.

mavaddat commented on .github/scripts/check_and_update_jdk.py:
```diff
@@ -0,0 +1,87 @@
+            f.write(latest_jdk)
+            
+        # Replace the current supported JDK version with the latest JDK version
+        readme = re.sub(readme_ver_pattern, latest_jdk, readme)
```
Right. I forgot about that one.

mavaddat commented:
> Ultimately it would depend on the support being merged into JDT-LS as you've discovered. And those depend on support in JDT Core (eg. [21](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/890), [22](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1397)).

Do we need JDT Core to add this as a feature? It seemed to me that we can track the tests that are already published in the CI:

| Num | Package | Class | Method | Latest test result |
| --- | --- | --- | --- | --- |
| 1 | `correction` | `ModifierCorrectionsQuickFixTest` | `testAddSealedMissingClassModifierProposal` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.correction/ModifierCorrectionsQuickFixTest/testAddSealedMissingClassModifierProposal/api/python) |
| 2 | `correction` | `ModifierCorrectionsQuickFixTest` | `testAddSealedAsDirectSuperClass` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.correction/ModifierCorrectionsQuickFixTest/testAddSealedAsDirectSuperClass/api/python) |
| 3 | `correction` | `ModifierCorrectionsQuickFixTest` | `testAddPermitsToDirectSuperClass` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.correction/ModifierCorrectionsQuickFixTest/testAddPermitsToDirectSuperClass/api/python) |
| 4 | `correction` | `UnresolvedTypesQuickFixTest` | `testTypeInSealedTypeDeclaration` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.correction/UnresolvedTypesQuickFixTest/testTypeInSealedTypeDeclaration/api/python) |
| 5 | `managers` | `EclipseProjectImporterTest` | `testPreviewFeaturesDisabledByDefault` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.managers/EclipseProjectImporterTest/testPreviewFeaturesDisabledByDefault/api/python) |
| 6 | `managers` | `InvisibleProjectImporterTest` | `testPreviewFeaturesEnabledByDefault` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.managers/InvisibleProjectImporterTest/testPreviewFeaturesEnabledByDefault/api/python) |
| 7 | `managers` | `MavenProjectImporterTest` | `testJava21Project` | [Link](https://ci.eclipse.org/ls/job/jdt-ls-master/lastCompletedBuild/testReport/org.eclipse.jdt.ls.core.internal.managers/MavenProjectImporterTest/testJava21Project/api/python) |

rgrunber commented:
No, I just wanted to explain that the changes that go into JDT-LS that are tracked in the script by certain test cases depend on JDT Core supporting that particular Java version.

mavaddat commented on .github/workflows/bump-jdk.yml:
```diff
@@ -0,0 +1,49 @@
+    env:
+        GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    steps:
+      - name: Checkout code
```
Fixed with [bf5a63c](https://github.com/mavaddat/vscode-java/commit/bf5a63ccc1c5cf8e99f777b141b02ddd4d01735e)

rgrunber commented on .github/scripts/check_and_update_jdk.py:
```diff
@@ -0,0 +1,92 @@
+            package = json.load(f)
+
+        # Add the latest JDK version to the java.configuration.runtimes array
+        package['contributes']['configuration']['properties']['java.configuration.runtimes']['items']['enum'].append(f'JavaSE-{latest_jdk}')
```
Since b1d08312b3d986368afbbe56e9271b4f9d6758fc , we took advantage of [settings groups](https://code.visualstudio.com/docs/getstarted/settings#_settings-groups), which made the configuration object contain a list of the groups, each containing a properties objects. This just needs a small change :

`next(filter(lambda e: e['id'] == "java-jdks", package['contributes']['configuration']))['properties']['java.configuration.runtimes']['items']['properties']['name']['enum']`

The 6th element in the configuration list should be the "Installed JDKs" settings group.

mavaddat replied:
Yes! I have updated the code to reflect your suggestion, @rgrunber .

mavaddat replied:
Updated with 17737b1cc94b1b36a324bacd660f2e94e9efde9f

rgrunber reviewed (commented):
Apologies for the delay. Tried this out and it looks good. Just a small change I noticed is needed to adjust to some changes we've made since.

If you don't have time to adjust it, let me know and I could take over as it seems to be in working order.

rgrunber reviewed (approved):
Based on my last attempt, this looks ready to merge. We're just preparing a release, and after that, we can add this in.

rgrunber commented:
@mavaddat , Just a heads-up the script didn't work for https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3111. Apparently the regex won't catch just `22` at https://javadl-esd-secure.oracle.com/update/baseline.version .

Also, not sure why I didn't catch this before, but the checks in the script, from the table above, that confirm JDT-LS has new version support and that reference "Sealed" should not be used. Sealed classes were finalized in Java 17 but for some reason it probably looked like their version should continue to be bumped. Probably just the 3 importer related tests are enough.

mavaddat commented:
> @mavaddat , Just a heads-up the script didn't work for [eclipse-jdtls/eclipse.jdt.ls#3111](https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3111). Apparently the regex won't catch just `22` at https://javadl-esd-secure.oracle.com/update/baseline.version .

Right. I see why. I did not know that major releases could have no minor semver part. I have corrected this in the 49e3b48eb6280480c92699b2bd94f81a0594faca commit.

> Also, not sure why I didn't catch this before, but the checks in the script, from the table above, that confirm JDT-LS has new version support and that reference "Sealed" should not be used. Sealed classes were finalized in Java 17 but for some reason it probably looked like their version should continue to be bumped. Probably just the 3 importer related tests are enough.

Oh, that's too bad. I kind of liked the loop I created to check for those tests. 😅 I have removed them nevertheless on 49e3b48eb6280480c92699b2bd94f81a0594faca.

rgrunber commented:
Yeah that format looked odd to me as well, but feel free to propose a PR for it.

It also published this a few hours ago.. I suspect because of  :
```
 Latest JDK version: 21
Current supported JDK version: 22
New JDK version detected: 21
```

at https://github.com/redhat-developer/vscode-java/pull/3547.
